apps/backend/src/app/modules/session/run/run-processor.class.ts

```diff
@@ -1,97 +1,130 @@
+  readonly zones: MapZones;
+  readonly replayHeader: ReplayFile.ReplayHeader;
+  readonly splits: RunSplits;
+
+  private static readonly logger = new Logger('Run Processor');
+
+  // Could pull out to config file/env if needed
+  static readonly Constants = {
+    // 10 seconds for the timer stage -> end record -> submit,
+    // then we add a second for every minute in the replay so longer replays
+    // have more time to submit, up to a max of 30s.
+    // Allowed time diff between recording finish and backend validation reached
+    AllowedSubmitDelayBase: 10_000,
+    AllowedSubmitDelayIncrement: 1000,
+    AllowedSubmitDelayMax: 30_000,
+
+    // Allowed time diff between split time and backend timestamp created
+    AllowedTimestampDelay: 5000
+  };
+
+  private constructor(
+    buffer: Buffer,
+    session: CompletedRunSession,
+    user: User
+  ) {
+    try {
+      this.buffer = buffer;
+      this.session = session;
+      this.user = user;
+      this.zones = JSON.parse(session.mmap.currentVersion.zones);
+      this.replayHeader = ReplayFile.Reader.readHeader(this.buffer);
+      this.splits = ReplayFile.Reader.readRunSplits(this.buffer);
+    } catch {
+      throw new RunValidationError(ErrorType.BAD_REPLAY_FILE);
+    }
   }
 
-  validateTimestamps() {
+  /** @throws {RunValidationError} */
+  static parse(buffer: Buffer, session: CompletedRunSession, user: User) {
+    return new RunProcessor(buffer, session, user);
+  }
+
+  /** @throws {RunValidationError} */
+  validateSessionTimestamps() {
+    const { timestamps, trackType, trackNum } = this.session;
+
     // Not giving specific reasons for throwing - if this ever happens they
     // an update timed out, the map zones are buggy, or someone's submitting
     // something nefarious. There's better ways to warn the user if they time
     // out, ideally as soon as it happens. If it's something nefarious don't
     // help them out with detailed errors.
     //
-    // Note that that currently, the timestamps do NOT include hitting the end
-    // zone. The game is stupid and can't send a multipart form, and needs to
-    // send a replay file, so we determine the final time by parsing the replay,
-    // rather than a timestamp. This will probably change in the future!
+    // Note that timestamps do NOT include hitting the end zone - hitting the
+    // end zone calls the /end endpoint with the replay, and we parse the replay
+    // header to determine the final time.
 
-    if (this.timestamps.length === 0) {
+    if (timestamps.length === 0) {
       throw new RunValidationError(ErrorType.BAD_TIMESTAMPS);
     }
 
-    // Check time is always increasing
-    for (let i = 1; i < this.timestamps.length; i++) {
-      if (this.timestamps[i].time < this.timestamps[i - 1].time) {
-        throw new RunValidationError(ErrorType.BAD_TIMESTAMPS);
-      }
+    // First minorNum is always 1 (a segment start), regardless of track type
+    if (timestamps[0].minorNum !== 1) {
```

Comment:
Is `timestamps` guaranteed to be ordered at this point, or could it still be out of order depending on network conditions? There's a few places where `timestamp` elements are accessed according to an assumed order.

Reply:
Will refactor as discussed on Discord, but need to do when more awake (next couple days)